src/opt/LowerISPCIntrinsics.cpp

```diff
@@ -0,0 +1,327 @@
+    } else if (opName == "umax") {
+        op = llvm::AtomicRMWInst::BinOp::UMax;
+    } else if (opName == "umin") {
+        op = llvm::AtomicRMWInst::BinOp::UMin;
+    } else if (opName == "fadd") {
+        op = llvm::AtomicRMWInst::BinOp::FAdd;
+    } else if (opName == "fsub") {
+        op = llvm::AtomicRMWInst::BinOp::FSub;
+    } else if (opName == "fmax") {
+        op = llvm::AtomicRMWInst::BinOp::FMax;
+    } else if (opName == "fmin") {
+        op = llvm::AtomicRMWInst::BinOp::FMin;
+    }
+    Assert(op != llvm::AtomicRMWInst::BinOp::BAD_BINOP);
+
+    llvm::AtomicOrdering ordering = lSetMemoryOrdering(memoryOrdering);
+    Assert(ordering != llvm::AtomicOrdering::NotAtomic);
+
+    return builder.CreateAtomicRMW(op, P, V, llvm::MaybeAlign(), ordering);
+}
+
+static llvm::Value *lLowerCmpXchgIntrinsic(llvm::CallInst *CI) {
+    // generate cmpxchg instruction fetching success ordering and failure
+    // ordering from intrinsic name
+    // llvm.ispc.cmpxchg.<successOrdering>.<failureOrdering>.<type>
+    llvm::IRBuilder<> builder(CI);
+
+    llvm::Value *P = CI->getArgOperand(0);
+    llvm::Value *C = CI->getArgOperand(1);
+    llvm::Value *N = CI->getArgOperand(2);
+
+    std::string prefix = CI->getCalledFunction()->getName().str();
+    prefix = prefix.substr(0, prefix.find_last_of('.'));
+    std::string successOrdering = prefix.substr(prefix.find_last_of('.') + 1);
+    prefix = prefix.substr(0, prefix.find_last_of('.'));
+    std::string failureOrdering = prefix.substr(prefix.find_last_of('.') + 1);
+
+    llvm::AtomicOrdering SO = lSetMemoryOrdering(successOrdering);
+    llvm::AtomicOrdering FO = lSetMemoryOrdering(failureOrdering);
+    Assert(SO != llvm::AtomicOrdering::NotAtomic && FO != llvm::AtomicOrdering::NotAtomic);
+
+    // cmpxchg returns a struct with two values, we only need the first one
+    llvm::Value *result = builder.CreateAtomicCmpXchg(P, C, N, llvm::MaybeAlign(), SO, FO);
+    return builder.CreateExtractValue(result, 0);
+}
+
+static llvm::Value *lLowerSelectIntrinsic(llvm::CallInst *CI) {
+    // generate select instruction
+    llvm::IRBuilder<> builder(CI);
+
+    llvm::Value *C = CI->getArgOperand(0);
+    llvm::Value *T = CI->getArgOperand(1);
+    llvm::Value *F = CI->getArgOperand(2);
+
+    // when C is not a vector of i1, we need to truncate it to i1
+    // This is ugly hack due to the fact that varying bool is represented as
+    // vector of i32/i16/i8 for some targets
+    llvm::VectorType *VT = llvm::dyn_cast<llvm::VectorType>(C->getType());
+    if (VT) {
+        // check if the vector element type is not i1
+        llvm::Type *ET = VT->getElementType();
+        if (!ET->isIntegerTy(1)) {
+            // truncate vector of i32/i16/i8 to vector of i1
+            llvm::Type *i1 = llvm::IntegerType::get(builder.getContext(), 1);
+            llvm::Type *newVT = llvm::VectorType::get(i1, lGetVecNumElements(C), false);
+            C = builder.CreateTrunc(C, newVT);
+        }
+    }
```

Comment:
Not mandatory, but this can be extracted to separate function.

Reply:
Done